Sources/SmokeDynamoDB/InMemoryDynamoDBCompositePrimaryKeyTableStore.swift

```diff
@@ -0,0 +1,427 @@
+
+    internal var store: [String: [String: PolymorphicOperationReturnTypeConvertable]] = [:]
+    internal let accessQueue = DispatchQueue(
+        label: "com.amazon.SmokeDynamoDB.InMemoryDynamoDBCompositePrimaryKeyTable.accessQueue",
+        target: DispatchQueue.global())
+    
+    internal let executeItemFilter: ExecuteItemFilterType?
+
+    init(executeItemFilter: ExecuteItemFilterType? = nil) {
+        self.executeItemFilter = executeItemFilter
+    }
+    
+    func getStore(eventLoop: EventLoop) -> EventLoopFuture<[String: [String: PolymorphicOperationReturnTypeConvertable]]> {
+        let promise = eventLoop.makePromise(of: [String: [String: PolymorphicOperationReturnTypeConvertable]].self)
+        
+        accessQueue.async {
+            promise.succeed(self.store)
+        }
+        
+        return promise.futureResult
+    }
+
+    func insertItem<AttributesType, ItemType>(_ item: TypedDatabaseItem<AttributesType, ItemType>,
+                                              eventLoop: EventLoop) -> EventLoopFuture<Void> {
+        let promise = eventLoop.makePromise(of: Void.self)
+        
+        accessQueue.async {
+            let partition = self.store[item.compositePrimaryKey.partitionKey]
+
+            // if there is already a partition
+            var updatedPartition: [String: PolymorphicOperationReturnTypeConvertable]
+            if let partition = partition {
+                updatedPartition = partition
+
+                // if the row already exists
+                if partition[item.compositePrimaryKey.sortKey] != nil {
+                    let error = SmokeDynamoDBError.conditionalCheckFailed(partitionKey: item.compositePrimaryKey.partitionKey,
+                                                                          sortKey: item.compositePrimaryKey.sortKey,
+                                                                          message: "Row already exists.")
+                    
+                    promise.fail(error)
+                    return
+                }
+
+                updatedPartition[item.compositePrimaryKey.sortKey] = item
+            } else {
+                updatedPartition = [item.compositePrimaryKey.sortKey: item]
+            }
+
+            self.store[item.compositePrimaryKey.partitionKey] = updatedPartition
+            promise.succeed(())
+        }
+        
+        return promise.futureResult
+    }
+
+    func clobberItem<AttributesType, ItemType>(_ item: TypedDatabaseItem<AttributesType, ItemType>,
+                                               eventLoop: EventLoop) -> EventLoopFuture<Void> {
+        let promise = eventLoop.makePromise(of: Void.self)
+        
+        accessQueue.async {
+            let partition = self.store[item.compositePrimaryKey.partitionKey]
+
+            // if there is already a partition
+            var updatedPartition: [String: PolymorphicOperationReturnTypeConvertable]
+            if let partition = partition {
```

Comment:
nit: can we just do this `updatedPartition = partition` and remove the line under this?

Reply:
It would make it slightly awkward as updatedPartition would become optional and you would then having to unwrap it to update it on the next line. I have left the code as it is but let me know if you still have concerns.

Reply:
Ah got it, makes sense.